### PR TITLE
Add postgresql history sync

### PR DESCRIPTION
**Pull request by enjuthulung**
CU-25ch1w[closed]

### Summary

- Added PostgreSQL history sync

### CODE REVIEW

**Comment by NubeDev** [comment]
Can you please make it that user can enter remote from the `config.json` database details not just localhost

https://github.com/NubeIO/rubix-point-server/pull/227/files#diff-432b65341693b39cec1910af65ba4d2613b52dde3322838fa6ee20cc72fe0ffcR55

File: tests/setting_dump.py [hunk 1]
```diff
@@ -12,7 +12,8 @@
         "mqtt": true,
         "histories": true,
         "cleaner": true,
-        "history_sync": true
+        "history_sync": true,
```
Comment (RaiBnod) [inline tests/setting_dump.py:15]: Change it into `history_sync_influxdb`

File: tests/setting_dump.py [hunk 2]
```diff
@@ -12,7 +12,8 @@
...
+        "history_sync_postgres": true
```
Comment (RaiBnod) [inline tests/setting_dump.py:16]: Add your config example for postgres.

**Review by RaiBnod — CHANGES_REQUESTED** [review]

File: src/services/histories/sync/postgresql.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+                    'table': self.config.table_name,
+                    'time': point_store_history.ts_value,
+                    'details': data_details,
+                    'fields': point_store_history
+                }
+                store.append(row)
+        if len(store):
+            logger.debug(f"Storing: {store}")
+            self.__client.write_points(store)
+            logger.info(f'Stored {len(store)} rows on {self.config.table_name} table')
```
Comment (RaiBnod) [inline src/services/histories/sync/postgresql.py:108]: These are done for bulk insertion on InfluxDB, it should show exceptions but exceptions are handled so app is not crashing.

File: src/services/histories/sync/postgresql.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                query = f'INSERT INTO {self.config.table_name} (id,point_uuid,value,value_original,value_raw,fault,' \
+                        f'fault_message,ts_value,ts_fault) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)'
+                data = (point_store_history.id, point_store_history.point_uuid, point_store_history.value,
+                        point_store_history.value_original, point_store_history.value_raw, point_store_history.fault,
+                        point_store_history.fault_message, point_store_history.ts_value, point_store_history.ts_fault)
+                with self.__client:
+                    with self.__client.cursor() as curs:
+                        try:
+                            curs.execute(query_details, data_details)
+                            curs.execute(query, data)
```
Comment (RaiBnod) [inline src/services/histories/sync/postgresql.py:95]: Can you see on bulk insertion as like this:

```sql
INSERT INTO films (code, title, did, date_prod, kind) VALUES
    ('B6717', 'Tampopo', 110, '1985-02-10', 'Comedy'),
    ('HG120', 'The Dinner Game', 140, DEFAULT, 'Comedy');
```

File: src/services/histories/sync/postgresql.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    def _create_table_if_not_exists(self):
+        query_details = f'CREATE TABLE IF NOT EXISTS {self.config.table_name}_details (point_uuid VARCHAR' \
+                        f' PRIMARY KEY,client_id VARCHAR(80),site_id VARCHAR(80),device_id VARCHAR(80),' \
+                        f'point_name VARCHAR(80), device_name VARCHAR(80),network_name VARCHAR(80),' \
+                        f'driver VARCHAR(80));'
+        query = f'CREATE TABLE IF NOT EXISTS {self.config.table_name} (id INTEGER PRIMARY KEY,point_uuid VARCHAR,' \
+                f'value NUMERIC,value_original NUMERIC,value_raw VARCHAR,fault BOOLEAN,fault_message VARCHAR,' \
+                f'ts_value TIMESTAMP,ts_fault TIMESTAMP,CONSTRAINT fk_{self.config.table_name}_details' \
+                f' FOREIGN KEY(point_uuid) REFERENCES' \
+                f' {self.config.table_name}_details(point_uuid));'
```
Comment (RaiBnod) [inline src/services/histories/sync/postgresql.py:121]: Space on end instead of on starting would have been good formatting.

**Review by RaiBnod — CHANGES_REQUESTED** [review]

File: src/services/histories/sync/postgresql.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                data_history.append(history)
+                row = {
+                    'table': self.config.table_name,
+                    'time': point_store_history.ts_value,
+                    'details': detail,
+                    'fields': history
+                }
+                store.append(row)
+        if len(store):
+            logger.debug(f"Storing: {store}")
```
Comment (RaiBnod) [inline src/services/histories/sync/postgresql.py:101]: No, needs of have that row and store. Just use:
```python
        if len(data_history):
            logger.info(f"Storing data_detail: {data_detail}")
            logger.info(f"Storing data_history: {data_history}")
```

**Review by RaiBnod — CHANGES_REQUESTED** [review]
Also there is two `device_name` one from `wires_plat` & another from `history point`. Make `edge_device_uuid` & `edge_device_name` for `history_point` and leave `device_name` & `device_id`. Make sure this changes should be for both `InfluxDB` tags and `history_details`.

File: config/config.example.json [hunk 7]
```diff
@@ -23,7 +24,20 @@
...
+        "host": "0.0.0.0",
+        "port": 5432,
+        "dbname": "db",
+        "user": "user",
+        "password": "password",
+        "ssl_mode": "allow",
+        "connect_timeout": 5,
+        "timer": 1,
+        "table_name": "history",
+        "attempt_reconnect_secs": 5
```
Comment (RaiBnod) [inline config/config.example.json:40]: Make same format as others.

**Review by RaiBnod — APPROVED** [review]

**Pull request merged**